ddl/ddl_api.go

```diff
@@ -1096,6 +1096,68 @@ func checkConstraintNames(constraints []*ast.Constraint) error {
 	return nil
 }
 
+func setTableAutoRandomBits(tbInfo *model.TableInfo, colDefs []*ast.ColumnDef) error {
+	allowAutoRandom := config.GetGlobalConfig().Experimental.AllowAutoRandom
+	pkColName := tbInfo.GetPkName()
+	return forEachColWithAutoRandomOpt(colDefs, func(col *ast.ColumnDef, autoRandomBits uint64) error {
+		if !allowAutoRandom {
+			return ErrInvalidAutoRandom.GenWithStackByArgs(autoid.AutoRandomExperimentalDisabledErrMsg)
+		}
+
+		if tbInfo.PKIsHandle && col.Name.Name.L == pkColName.L {
+			if containsColumnOption(col, ast.ColumnOptionAutoIncrement) {
+				return ErrInvalidAutoRandom.GenWithStackByArgs(autoid.AutoRandomIncompatibleWithAutoIncErrMsg)
+			}
+			if containsColumnOption(col, ast.ColumnOptionDefaultValue) {
+				return ErrInvalidAutoRandom.GenWithStackByArgs(autoid.AutoRandomIncompatibleWithDefaultValueErrMsg)
+			}
+			fieldTypeBitsLength := uint64(mysql.DefaultLengthOfMysqlTypes[col.Tp.Tp] * 8)
+			if autoRandomBits == 0 {
```

Comment:
Why not check it in `convertAutoRandomBitsToUnsigned` on line1146 directly?

Reply:
Because `AutoRandomBits == 0` is not an error in some case. 

`convertAutoRandomBitsToUnsigned` is also used in `checkAutoRandom`, which in turn used in `ModifyColumn`. For example, we have the schema
```
create table t (a int);
``` 
The `AutoRandomBits` in `TableInfo` is 0. 

When we try to run `alter table t modify column a int`, the `specNewColumn` here is a __valid__ 0.